ext/openssl/openssl.c

```diff
@@ -5309,48 +5337,717 @@ PHP_FUNCTION(openssl_pkcs7_decrypt)
+	}
+
+	cert = php_openssl_x509_from_zval(zcert, 0, &certresource);
+	if (cert == NULL) {
+		php_error_docref(NULL, E_WARNING, "Error getting cert");
+		goto clean_exit;
+	}
+
+	if (php_openssl_open_base_dir_chk(infilename) || php_openssl_open_base_dir_chk(outfilename)) {
+		goto clean_exit;
+	}
+
+	if ((encoding & ENCODING_SMIME) && (flags & CMS_DETACHED)) {
+		php_error_docref(NULL,
+				 E_WARNING, "detached signatures not possible with S/MIME encoding");
+		goto clean_exit;
+	}
+
+        /* This flag is set because of what appears to be a bug in
+	 * SMIME_write_CMS.  It is perfectly fine because we are going
+	 * to finalize directly below.
+	 */
+	flags |= CMS_STREAM;
+	
+	infile = BIO_new_file(infilename, PHP_OPENSSL_BIO_MODE_R(flags));
+	if (infile == NULL) {
+		php_openssl_store_errors();
+		php_error_docref(NULL, E_WARNING, "Error opening input file %s!", infilename);
+		goto clean_exit;
+	}
+
+	outfile = BIO_new_file(outfilename, PHP_OPENSSL_BIO_MODE_W(CMS_BINARY));
+	if (outfile == NULL) {
+		php_openssl_store_errors();
+		php_error_docref(NULL, E_WARNING, "Error opening output file %s!", outfilename);
+		goto clean_exit;
+	}
+
+	cms = CMS_sign(cert, privkey, others, infile, (unsigned int)flags);
+	if (cms == NULL) {
+		php_openssl_store_errors();
+		php_error_docref(NULL, E_WARNING, "Error creating CMS structure!");
+		goto clean_exit;
+	}
+	if (BIO_reset(infile) != 0) {
+		php_openssl_store_errors();
+		goto clean_exit;
+	}
+	
+	/* tack on extra headers */
+	if (zheaders && encoding == ENCODING_SMIME ) {
+		int ret;
+
+		ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(zheaders), strindex, hval) {
+			zend_string *str = zval_try_get_string(hval);
+			if (UNEXPECTED(!str)) {
+				goto clean_exit;
+			}
+			if (strindex) {
+				ret = BIO_printf(outfile, "%s: %s\n", ZSTR_VAL(strindex), ZSTR_VAL(str));
+			} else {
+				ret = BIO_printf(outfile, "%s\n", ZSTR_VAL(str));
+			}
+			
+			zend_string_release(str);
+			if (ret < 0) {
+				php_openssl_store_errors();
+			}
+		} ZEND_HASH_FOREACH_END();
+	}
+	/* writing the signed data depends on the encoding */
+	switch ( encoding ) {
+		case ENCODING_SMIME:
+			if (!SMIME_write_CMS(outfile, cms, infile, (int)flags)) {
+	      			php_openssl_store_errors();
+					goto clean_exit;
+			}
+			break;
+		case ENCODING_DER:
+		  	if (CMS_final(cms,infile,NULL, (unsigned int) flags) != 1) {
+			  	php_openssl_store_errors();
+	      			goto clean_exit;
+			}
+			if (i2d_CMS_bio(outfile,cms) != 1) {
+			  	php_openssl_store_errors();
+				goto clean_exit;
+			}
+			break;
+	 	case ENCODING_PEM:
+		  	if (CMS_final(cms,infile,NULL, (unsigned int) flags) != 1) {
+			  	php_openssl_store_errors();
+				goto clean_exit;
+			}
+			if (PEM_write_bio_CMS(outfile, cms) == 0) {
+			  	php_openssl_store_errors();
+				goto clean_exit;
+			}
+			break;
+	  	default:
+		  	php_error_docref(NULL, E_WARNING, "unknown OPENSSL encoding");
+			goto clean_exit;
+	}
+	RETVAL_TRUE;
+
+clean_exit:
+	if (cms) {
```

Comment:
passing NULL to free functions should be fine so no need to check it here and below. It would be inconsistent with other functions.

Reply:
Ok.  I will do this with most of the free()s.  There are, I think, two that may require special handling.

Reply:
FYI I tried this and found that I wasn't passing Appveyor CI.

Reply:
I am still working on this.  BIO_free() is documented to do nothing if the argument is a NULL.  This is not the case with X509_STORE_free() or CMS_ContentInfo_free().

Reply:
Ok BIO_free is still outstanding, right?

Reply:
Right.  Missed one function.  Corrected.